Add GE limits to Grand Exchange Plugin Search

Asymons opened the pull request:
Fixes https://github.com/runelite/runelite/issues/4533

Preview:
![2018-09-23_19-37-48](https://user-images.githubusercontent.com/5115805/45931102-2cddba00-bf69-11e8-9063-663604550a7b.png)

MagicfTail commented:
Do `Fixes #4533` and it will auto close the issue

MagicfTail commented on runelite-client/src/main/java/net/runelite/client/game/ItemManager.java:
```diff
@@ -165,6 +179,25 @@ private void loadPrices()
+		{
+			log.warn("error loading limits!", e);
+		}
+
```
Remove empty line

MagicfTail commented on runelite-client/src/main/resources/net/runelite/client/plugins/grandexchange/ge_limits.json:
```diff
@@ -0,0 +1 @@
+{"0":-1,"1":-1,"2":7000,"3":-1,"4":-1,"5":-1,"6":-1,"8":-1,"10":-1,"12":-1,"14":-1,"15":-1,"16":-1,"17":-1,"18":-1,"19":-1,"20":-1,"21":-1,"22":-1,"23":-1,"24":-1,"25":-1,"26":-1,"27":-1,"28":-1,"30":-1,"32":-1,"33":-1,"35":-1,"36":-1,"38":-1,"39":-1,"40":-1,"41":-1,"42":-1,"43":-1,"44":-1,"45":-1,"46":-1,"47":-1,"48":-1,"50":-1,"52":7000,"53":7000,"54":-1,"56":-1,"58":-1,"60":-1,"62":-1,"64":100 ...
```
Make this more readable

MagicfTail commented on runelite-client/src/main/java/net/runelite/client/game/ItemManager.java:
```diff
@@ -328,4 +361,24 @@ public BufferedImage getItemOutline(final int itemId, final int itemQuantity, fi
+	 */
+	public int getItemGELimit(int itemID)
+	{
+		int limitQuantity = 0;
```
Empty line before if/for

Asymons commented:
made changes

Asymons commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java:
```diff
@@ -133,6 +135,13 @@ public void mouseReleased(MouseEvent e)
 
+		// GE Limit
+		JLabel geLimitLabel = new JLabel();
+		String limitLabelText = geItemLimit == -1 ? "No limit found" : StackFormatter.formatNumber(geItemLimit) + " item limit";
```
need to update this to 0 or null

Asymons replied:
fixed

Asymons commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java:
```diff
@@ -133,6 +135,13 @@ public void mouseReleased(MouseEvent e)
 		rightPanel.add(haPriceLabel);
 
+		// GE Limit
+		JLabel geLimitLabel = new JLabel();
```
I'll add that into this PR

Asymons commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java:
```diff
@@ -330,4 +343,13 @@ public void onGameTick(GameTick event)
+	private void loadGELimits()
+	{
+		InputStream geLimitData = GrandExchangeSearchPanel.class.getResourceAsStream("/ge_limits.json");
+		itemGELimits = gson.fromJson(new InputStreamReader(geLimitData), new TypeToken<Map<Integer,Integer>>(){}.getType());
```
TypeToken seems to be necessary. Occurrences exist in the project when dealing with data structures (like list) which have generics required

Asymons commented:
changes should be finished now

deathbeam commented on runelite-client/src/main/java/net/runelite/client/game/ItemManager.java:
```diff
@@ -84,6 +84,7 @@
 	private final LoadingCache<Integer, ItemComposition> itemCompositions;
 	private final LoadingCache<OutlineKey, BufferedImage> itemOutlines;
 
+
```
Remove this empty line

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java:
```diff
@@ -36,21 +36,23 @@
  * This panel displays an individual item result in the
  * Grand Exchange search plugin.
  */
+@Slf4j
```
Looks like you dont need this

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItems.java:
```diff
@@ -35,13 +35,15 @@
+	private final int geItemLimit;
 
-	GrandExchangeItems(AsyncBufferedImage icon, String name, int itemId, int gePrice, double haPrice)
+	GrandExchangeItems(AsyncBufferedImage icon, String name, int itemId, int gePrice, double haPrice, int geItemLimit)
```
Wait why is this constructor even here. Just remove it, the `@Data` annotation already creates constructor. and probably change `@Data` to `@Value` when at it

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeSearchPanel.java:
```diff
@@ -283,4 +288,9 @@ private void processResult(SearchResult result, String lookup, boolean exactMatc
 		});
 	}
 
+	public 	void setItemGELimits(Map<Integer, Integer> itemGELimits)
```
Lombok `@Setter` maybe?

Asymons commented:
I think it looks better with the extra row as the information is a consistent top down read. Probably best to get more feedback though about this.

deathbeam commented:
What about using item quantity as buy limit indicator? (e.g the top-left corner of item)

Asymons commented:
Implemented more changes, feel free to review. Implemented most of the suggestions off discord.
Screenshots:
With Limit in GE:
![new-limit](https://user-images.githubusercontent.com/1833964/44176722-f46cd580-a0b8-11e8-9b3e-9d3930b8da91.PNG)
Without Limit in GE:
![new-limit-without](https://user-images.githubusercontent.com/1833964/44176729-fc2c7a00-a0b8-11e8-8a9f-bcd5a37ea8a7.png)

psikoi commented:
Please add some right margin to the limits, looks too close to the edge of the container

Asymons commented:
added bigger margins

Asymons commented:
![capture](https://user-images.githubusercontent.com/1833964/44182436-f5136500-a0d4-11e8-9455-5420344dcc57.PNG)

deathbeam commented:
That "Buy limit:" string is way too long. What about **&uarr;15,000**

Asymons commented:
Changed the "Buy Limit: " string to "Limit: ". Seems to save a ton of space.
![image](https://user-images.githubusercontent.com/1833964/44209060-443ab380-a130-11e8-9eca-750520529dbc.png)

psikoi commented:
Ah nice, much better :D

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java:
```diff
@@ -330,4 +343,15 @@ public void onGameTick(GameTick event)
 			}
 		});
 	}
+
```
Remove this empty line

Asymons replied:
😶 I keep missing these Ahaha good catch

Asymons replied:
😶 I keep missing these Ahaha good catch

Asymons replied:
😶 I keep missing these Ahaha good catch

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java:
```diff
@@ -104,7 +105,7 @@ public void mouseReleased(MouseEvent e)
 		// Item details panel
 		JPanel rightPanel = new JPanel(new GridLayout(3, 1));
-		rightPanel.setBackground(ColorScheme.MEDIUM_GRAY_COLOR);
+		rightPanel.setBackground(ColorScheme.TRANSPARENT);
```
Are you sure you need to use transparent? Swing has usually bad performance with transparent components

Asymons replied:
Transparent is unnecessary on right panel but is necessary on the alchAndLimit panel unless we want to attach a second mouse listener to that child panel.

Asymons replied:
Transparent is unnecessary on right panel but is necessary on the alchAndLimit panel unless we want to attach a second mouse listener to that child panel.

Asymons commented:
Instead of using a mouse listener twice, I just made it so all JPanels were recorded in a list and colored their respective components.

Asymons commented:
Added more limits from the runescape wiki and the osbox db.

psikoi commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java:
```diff
@@ -127,15 +133,38 @@ public void mouseReleased(MouseEvent e)
 		add(rightPanel, BorderLayout.CENTER);
 	}
 
+	private void matchComponentBackground(JPanel panel, Color color)
```
Can't you just override the setBackground method?

psikoi replied:
And inside it, call the super.setBackground, along with that for loop, to apply the color to its children

Asymons replied:
I agree, I believe that would be cleaner.

Asymons replied:
I've looked into it again, the reason why I did it this way was that the loop of children only looks down a single level. All the children are considered components. This means I cannot recursively look for more components from each child since that method does not exist within the component class. Casting everything to a JPanel also gave me runtime erros for certain classes.

Asymons commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java:
```diff
@@ -317,6 +328,8 @@ public void onGameTick(GameTick event)
 		lastItem = itemId;
 
 		log.debug("Looking up item {}", itemId);
+		final int itemLimit = itemGELimits.getOrDefault(itemId, 0);
```
I'll be taking a look at it tonight, I remember them working w/ or w/o OSB prices enabled but will double check.

Asymons commented:
Will this be added to next week's milestone?

deathbeam commented:
Fixed remaining issues, updated preview:
![2018-09-23_19-30-19](https://user-images.githubusercontent.com/5115805/45931046-2995fe80-bf68-11e8-9db7-8c197d0cac94.png)
